### PR TITLE
Add Minio to the app install command

### LINKED ISSUE #56: App: Add Minio
## Expected Behaviour

Add Minio as an app with a default flag not to use persistent storage.

Generate the secret/access key or allow them via flag too.

Example command: https://github.com/alexellis/k3sup/blob/master/pkg/cmd/openfaas_app.go

## Possible Solution

Replicate approach from: https://github.com/openfaas-incubator/ofc-bootstrap/blob/master/scripts/install-minio.sh

### PR DESCRIPTION
**Pull request by burtonr**
## Description

This adds the ability to install Minio using the helm chart
There are flags added to allow the user to set the access key
and secret key as well as change the mode of installation

There is a safety net check to ensure that both access and
secret key are included if one is present

Signed-off-by: Burton Rheutan <rheutan7@gmail.com>

## Motivation and Context

- [X] I have raised an issue to propose this change ([required](https://github.com/openfaas/faas/blob/master/CONTRIBUTING.md))
Resolves #56 

## How Has This Been Tested?

Installed Minio app on local microk8s cluster

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [X] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [X] I've read the [CONTRIBUTION](https://github.com/openfaas/faas/blob/master/CONTRIBUTING.md) guide
- [X] I have signed-off my commits with `git commit -s`
- [ ] I have added tests to cover my changes.
- [X] All new and existing tests passed.

### CODE REVIEW

**Comment by alexellis** [comment]
Hi @burtonr thanks for this PR.

Since k3sup is about ease of use and working fast, I'd like to see the two credentials generated in the same way we do for OpenFaaS.

If the user provides an override for either, then don't generate.

**Comment by burtonr** [comment]
@alexellis Minio will generate those secrets automatically by default. However, after a bit of testing I found that the values generated by Minio were always the same...
So, I've updated the installation to generate the access and secret key if they are not provided. I've also updated the final output to describe how to access those secrets as well as port-forwarding to allow access.
Let me know if you'd like anything else

**Comment by alexellis** [comment]
@Waterdrips please can you review and try this out?

@burtonr thank you for updating, it needs a rebase now

File: pkg/cmd/minio_app.go [hunk 1]
```diff
@@ -0,0 +1,137 @@
...
+		if err != nil {
+			return err
+		}
+
+		overrides := map[string]string{}
+		accessKey, _ := minio.Flags().GetString("access-key")
+		secretKey, _ := minio.Flags().GetString("secret-key")
+
+		if len(accessKey) == 0 {
+			fmt.Printf("Access Key not provided, one will be generated for you")
```
Comment (Waterdrips) [inline pkg/cmd/minio_app.go]: no newline at the end of these prints - the lines get joined in output 

`Access Key not provided, one will be generated for youSecret Key not provided, one will be generated for youexec:  /Users/alistairhey/.k3sup/bin/helm template minio --name minio --namespace minio --output-dir /var/folders/y2/67nr0tqd10q35gdq84_mkz2h0000gn/T/charts/minio/rendered --values /var/folders/y2/67nr0tqd10q35gdq84_mkz2h0000gn/T/charts/minio/values.yaml  --set accessKey=#### --set secretKey=######
`

**Comment by burtonr** [comment]
Thanks for the feedback @Waterdrips :100: I'll get those things taken care of right away and be sure to test, and re-test, on a new clean cluster.

File: pkg/cmd/minio_app.go [hunk 2]
```diff
@@ -0,0 +1,137 @@
...
+		}
+
+		if err != nil {
+			return err
+		}
+
+		overrides["accessKey"] = accessKey
+		overrides["secretKey"] = secretKey
+
+		if dist, _ := minio.Flags().GetBool("distributed-mode"); dist {
```
Comment (alexellis) [inline pkg/cmd/minio_app.go]: Since the flags are modes, let's just use `distributed` instead?

Comment (burtonr) [reply pkg/cmd/minio_app.go]: Good point, updated to remove the "-mode" from the flag

**Comment by burtonr** [comment]
@Waterdrips Please could you try this again on your side?

I've taken the time to carefully clean my system (restart machine), clean the cluster (`microk8s.reset --remove-storage`), clean my local files (`rm -rf /tmp/charts`) and clean my environment (new terminal window). I've since run this about 5 times successfully. 

Please do let me know if there is still an issue. It is possible I am missing something on my side between tests that is allowing it to pass for me.

**Comment by blaise-sumo** [comment]
These are such thorough and constructive reviews that I find myself wanting
to learn golang!

File: pkg/cmd/minio_app.go [hunk 3]
```diff
@@ -0,0 +1,137 @@
...
+
+# Get the access and secret key to gain access to minio
+ACCESSKEY=$(kubectl get secret -n default minio -o jsonpath="{.data.accesskey}" | base64 --decode; echo)
+SECRETKEY=$(kubectl get secret -n default minio -o jsonpath="{.data.secretkey}" | base64 --decode; echo)
+
+# Get the Minio Client
+wget https://dl.min.io/client/mc/release/linux-amd64/mc \
+&& chmod +x mc \
+&& ./mc --help
+
```
Comment (alexellis) [inline pkg/cmd/minio_app.go]: Let's just add the `mc config host with above` command?

Comment (alexellis) [reply pkg/cmd/minio_app.go]: Send in a new PR?

**Comment by alexellis** [comment]
Rebase required on this

**Comment by burtonr** [comment]
@alexellis I've rebased into a single commit, so this is ready to go now. Thank you

**Comment by alexellis** [comment]
Couple of things I noticed:

* Missing the command to add the Minio server after downloading mc, that has been added
* The chart wouldn't deploy on k3s because it was requesting a PV - persistent claim, this is now off by default with an option to enable
* The instructions told the user to download a Linux binary, without looking at the client architecture, that's now dynamic.

> How Has This Been Tested?

Installed Minio app on local microk8s cluster

^ I do wonder if this worked for you? My cluster sat in Pending mode, but maybe microk8s includes a local volume provisioner out of the box?

Thanks for the contribution, hope you'll pick up another one soon 👍 

https://github.com/alexellis/k3sup/commit/9af477914199a92e4daa140bc3dad208745049c4

**Pull request merged**